### PR TITLE
Add highlights.scm query for Solidity

### PR DESCRIPTION
**Pull request by YongJieYongJie**
I noticed the [original issue](https://github.com/nvim-treesitter/nvim-treesitter/issues/1168) has been inactive for quite some time, and thought I might take a stab at adding basic highlighting support for Solidity.

Here's a preview of the highlighting:
image

Also managed to add the test and got the `./scripts/run_tests.sh query` passing for Solidity:
image

Do let me know if this is alright :bowing_man:

### CODE REVIEW

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -959,6 +959,15 @@ list.lalrpop = {
   maintainers = { "@traxys" },
 }
 
+list.solidity = {
+  install_info = {
+    url = "https://github.com/YongJieYongJie/tree-sitter-solidity",
+    branch = "with-generated-c-code",
```
Comment (theHamsta) [inline lua/nvim-treesitter/parsers.lua:965]: are you planning to maintain this branch?

File: queries/solidity/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+ "switch"
+ "case"
+ "default"
+ "break"
+ "continue"
+ "if"
+ "else"
+ "for"
+ "while"
+ "do"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:110]: `@repeat`

File: queries/solidity/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+ "function"
+ "var"
+ (constant)
+ (virtual)
+ (override_specifier)
+ (yul_leave)
+] @keyword
+
+(import_directive "as" @keyword)
+(import_directive "from" @keyword)
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:136]: both `@include`?

File: queries/solidity/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  "!="
+  "!=="
+  ">="
+  ">"
+  "!"
+  "~"
+  "-"
+  "+"
+  "delete"
+  "new"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:186]: `@keyword.operator` (has same highlight as `@operator` but allows more customization

File: tests/query/highlights/solidity/test.sol [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    // Calls winningProposal() function to get the index
+    // of the winner contained in the proposals array and then
+    // returns the name of the winner
+    function winnerName() external view
+            returns (bytes32 winnerName_)
+    {
+        winnerName_ = proposals[winningProposal()].name;
+    }
+}
```
Comment (theHamsta) [inline tests/query/highlights/solidity/test.sol:171]: thanks, for adding tests :+1: !

File: lua/nvim-treesitter/parsers.lua (hunk 1 above)
Comment (YongJieYongJie) [reply lua/nvim-treesitter/parsers.lua:965]: @theHamsta yes, I'll be maintaining this branch, keeping the generated C code updated with master (and upstream).

File: queries/solidity/highlights.scm (hunk 3 above)
Comment (YongJieYongJie) [reply queries/solidity/highlights.scm:136]: Yes, and also the `import` from the larger list of keywords above. Let me make the necessary changes.

File: queries/solidity/highlights.scm (hunk 4 above)
Comment (YongJieYongJie) [reply queries/solidity/highlights.scm:186]: Noted :+1:

**Comment by YongJieYongJie** [comment]
@theHamsta Thanks for the quick review,

I have made the following changes:
- use more appropriate/specific captures as per comments above,
- fixed the CI jobs that were failing because of invalid captures, and also
- improved highlighting of `type_name` containing punctuation (e.g., the brackets in the `bytes32[]` used to be highlighted in yellow too)
image

File: queries/solidity/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+ "external"
+ "pure"
+ "view"
+ "payable"
+ "modifier"
+ "returns"
+ "memory"
+ "storage"
+ "calldata"
+ "function"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:124]: `@keyword.function`

File: queries/solidity/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+ "switch"
+ "case"
+ "default"
+ "break"
+ "continue"
+ "if"
+ "else"
+ "try"
+ "catch"
+ "return"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:110]: `@keyword.return`

File: queries/solidity/highlights.scm [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+ "event"
+ "using"
+ "assembly"
+ "switch"
+ "case"
+ "default"
+ "break"
+ "continue"
+ "if"
+ "else"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:107]: `@conditional`

File: queries/solidity/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+ "try"
+ "catch"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:109]: `@exception`

File: queries/solidity/highlights.scm [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+(enum_value) @property
+
+
+; Keywords
+[
+ "pragma"
+ "contract"
+ "interface"
+ "library"
+ "is"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:95]: maybe `@keyword.operator`. Not sure, don't know the language.

File: queries/solidity/highlights.scm [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+ "struct"
+ "enum"
+ "event"
+ "using"
+ "assembly"
+ "switch"
+ "case"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:102]: `@conditional`

File: queries/solidity/highlights.scm [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+ "emit"
+ "public"
+ "internal"
+ "private"
+ "external"
+ "pure"
+ "view"
+ "payable"
+ "modifier"
+ "returns"
```
Comment (theHamsta) [inline queries/solidity/highlights.scm:120]: `@keyword.return`? Not sure.

File: queries/solidity/highlights.scm (hunk 7 above)
Comment (YongJieYongJie) [reply queries/solidity/highlights.scm]: Noted, thanks :pray:

File: queries/solidity/highlights.scm (hunk 11 above)
Comment (YongJieYongJie) [reply queries/solidity/highlights.scm]: Noted, thanks :pray: (and `default` as the default switch-case too)

**Comment by YongJieYongJie** [comment]
@theHamsta Thanks again for the quick review.

Apologies the missing out on the various captures (I had originally wanted to keep the highlights.scm query file more-or-less similar to upstream).

I have went through the various available captures in the [Contributing](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights) page and updated the file accordingly. Do have a look when you have time :pray:

**Review by theHamsta — APPROVED** [review]

**Comment by theHamsta** [comment]
Everything should be fine now. Thank you a lot!

**Pull request merged**